File: clipsync-win/ClipSync.Core/Net/StreamAssembler.cs
using System;
using System.Collections.Generic;
using System.Security.Cryptography;

namespace ClipSync.Net;

public enum StreamOutcome { Ok, Ignored, Dropped }
public readonly record struct StreamResult(StreamOutcome Outcome, string? Reason = null)
{
    public static readonly StreamResult Ok = new(StreamOutcome.Ok);
    public static StreamResult Ignore(string why) => new(StreamOutcome.Ignored, why);
    public static StreamResult Drop(string why) => new(StreamOutcome.Dropped, why);
}

/// Reassembles one connection's streamed ClipboardItem: the wire item is
/// parked, FileChunks fill a buffer per stream_id, FileEnd verifies, and
/// once every stream is done the item is rebuilt with those payloads
/// inline -- indistinguishable downstream from an item that arrived
/// inline (writers, loop-suppression hash, history). Pure state machine;
/// PeerConnection feeds it and logs the reasons it returns.
///
/// One pending item at a time: the sender is sequential, so a new parked
/// item means any unfinished older one is stale and is discarded.
public sealed class StreamAssembler
{
    private sealed class Slot
    {
        public required ClipFormat Format;
        public required byte[] Buffer;
        public int Received;
        public bool Done;
    }

    public static readonly TimeSpan IdleWindow = TimeSpan.FromSeconds(30);

    private ClipboardItem? _pending;
    private readonly Dictionary<ulong, Slot> _slots = new();
    private DateTime _lastProgress;

    public bool HasPending => _pending is not null;

    /// True if the item carries any stream_id format and so must be parked.
    public static bool NeedsAssembly(ClipboardItem item)
    {
        foreach (var f in item.Formats)
            if (f.Inline is null && f.StreamId is not null) return true;
        return false;
    }

    /// Park a wire item whose formats include stream_ids. Rejects (and
    /// parks nothing) if the declared sizes exceed the cap, so a peer can
    /// never make us allocate more than we would send ourselves.
    public StreamResult Park(ClipboardItem wire, DateTime now)
    {
        ulong total = 0;
        foreach (var f in wire.Formats)
        {
            var size = f.Inline is { } d ? (ulong)d.Length : f.Size;
            if (size > StreamPlanner.MaxItemBytes)
                return StreamResult.Drop($"declared {f.Mime} size {size} exceeds cap");
            total += size;
            if (total > StreamPlanner.MaxItemBytes)
                return StreamResult.Drop($"declared item size exceeds cap");
        }

        var replaced = _pending is not null;
        Reset();
        _pending = wire;
        foreach (var f in wire.Formats)
        {
            if (f.Inline is not null || f.StreamId is not { } id) continue;
            if (_slots.ContainsKey(id)) { Reset(); return StreamResult.Drop($"duplicate stream_id {id}"); }
            _slots[id] = new Slot { Format = f, Buffer = new byte[f.Size] };
        }
        _lastProgress = now;
        return replaced ? new StreamResult(StreamOutcome.Ok, "replaced an incomplete item") : StreamResult.Ok;
    }

    public StreamResult Chunk(ulong streamId, ulong offset, ReadOnlySpan<byte> data, DateTime now)
    {
        if (!_slots.TryGetValue(streamId, out var s)) return StreamResult.Ignore($"unknown stream_id {streamId}");
        if (s.Done) { Reset(); return StreamResult.Drop($"chunk after end for stream {streamId}"); }
        if (offset != (ulong)s.Received) { Reset(); return StreamResult.Drop($"offset {offset} != received {s.Received} for stream {streamId}"); }
        if ((ulong)s.Received + (ulong)data.Length > (ulong)s.Buffer.Length) { Reset(); return StreamResult.Drop($"chunk overruns declared size for stream {streamId}"); }
        data.CopyTo(s.Buffer.AsSpan(s.Received));
        s.Received += data.Length;
        _lastProgress = now;
        return StreamResult.Ok;
    }

    public StreamResult End(ulong streamId, ulong totalSize, ReadOnlySpan<byte> sha256, DateTime now)
    {
        if (!_slots.TryGetValue(streamId, out var s)) return StreamResult.Ignore($"unknown stream_id {streamId}");
        if (s.Done) { Reset(); return StreamResult.Drop($"duplicate end for stream {streamId}"); }
        if (totalSize != (ulong)s.Buffer.Length || s.Received != s.Buffer.Length)
        { Reset(); return StreamResult.Drop($"end size {totalSize}/{s.Received} != declared {s.Buffer.Length} for stream {streamId}"); }
        Span<byte> actual = stackalloc byte[32];
        SHA256.HashData(s.Buffer, actual);
        if (!actual.SequenceEqual(sha256)) { Reset(); return StreamResult.Drop($"hash mismatch for stream {streamId}"); }
        s.Done = true;
        _lastProgress = now;
        return StreamResult.Ok;
    }

    /// The materialized item once every stream has verified, else null.
    /// Consumes the pending state.
    public ClipboardItem? TakeCompleted()
    {
        if (_pending is null) return null;
        foreach (var s in _slots.Values) if (!s.Done) return null;
        var formats = new List<ClipFormat>(_pending.Formats.Count);
        foreach (var f in _pending.Formats)
        {
            if (f.Inline is null && f.StreamId is { } id && _slots.TryGetValue(id, out var s))
                formats.Add(new ClipFormat(f.Mime, (ulong)s.Buffer.Length, s.Buffer, null));
            else
                formats.Add(f);
        }
        var item = _pending with { Formats = formats };
        Reset();
        return item;
    }

    public bool IsStale(DateTime now) => _pending is not null && now - _lastProgress > IdleWindow;

    /// Drop the pending item if it has seen no progress within the window.
    /// Returns true if something was dropped.
    public bool DropStale(DateTime now)
    {
        if (!IsStale(now)) return false;
        Reset();
        return true;
    }

    public void Reset()
    {
        _pending = null;
        _slots.Clear();
    }
}

File: clipsync-mac/Sources/ClipSync/Net/StreamAssembler.swift
import Foundation
import Crypto

enum StreamOutcome: Equatable { case ok, ignored, dropped }
struct StreamResult: Equatable {
    let outcome: StreamOutcome
    let reason: String?
    static let ok = StreamResult(outcome: .ok, reason: nil)
    static func ignore(_ why: String) -> StreamResult { StreamResult(outcome: .ignored, reason: why) }
    static func drop(_ why: String) -> StreamResult { StreamResult(outcome: .dropped, reason: why) }
}

/// Reassembles one connection's streamed ClipboardItem: the wire item is
/// parked, FileChunks fill a buffer per stream_id, FileEnd verifies, and
/// once every stream is done the item is rebuilt with those payloads
/// inline -- indistinguishable downstream from an item that arrived
/// inline (writer, loop-suppression hash, Clipboard History). Pure state
/// machine; PeerConnection feeds it and logs the reasons it returns.
///
/// One pending item at a time: the sender is sequential, so a new parked
/// item means any unfinished older one is stale and is discarded.
///
/// Mirrors the Windows StreamAssembler.
final class StreamAssembler {
    private final class Slot {
        let format: ClipFormat
        var buffer: Data
        var received = 0
        var done = false
        init(format: ClipFormat) {
            self.format = format
            self.buffer = Data(count: Int(format.size))
        }
    }

    static let idleWindow: TimeInterval = 30

    private var pending: ClipboardItem?
    private var slots: [UInt64: Slot] = [:]
    private var lastProgress = Date.distantPast

    var hasPending: Bool { pending != nil }

    /// True if the item carries any stream_id format and so must be parked.
    static func needsAssembly(_ item: ClipboardItem) -> Bool {
        item.formats.contains { if case .stream = $0.payload { return true } else { return false } }
    }

    /// Park a wire item whose formats include stream_ids. Rejects (and
    /// parks nothing) if the declared sizes exceed the cap, so a peer can
    /// never make us allocate more than we would send ourselves.
    func park(_ wire: ClipboardItem, now: Date) -> StreamResult {
        var total: UInt64 = 0
        for f in wire.formats {
            let size: UInt64
            if case .inline(let d) = f.payload { size = UInt64(d.count) } else { size = f.size }
            if size > StreamPlanner.maxItemBytes {
                return .drop("declared \(f.mime) size \(size) exceeds cap")
            }
            total += size
            if total > StreamPlanner.maxItemBytes { return .drop("declared item size exceeds cap") }
        }

        let replaced = pending != nil
        reset()
        pending = wire
        for f in wire.formats {
            guard case .stream(let id) = f.payload else { continue }
            if slots[id] != nil { reset(); return .drop("duplicate stream_id \(id)") }
            slots[id] = Slot(format: f)
        }
        lastProgress = now
        return replaced ? StreamResult(outcome: .ok, reason: "replaced an incomplete item") : .ok
    }

    func chunk(streamId: UInt64, offset: UInt64, data: Data, now: Date) -> StreamResult {
        guard let s = slots[streamId] else { return .ignore("unknown stream_id \(streamId)") }
        if s.done { reset(); return .drop("chunk after end for stream \(streamId)") }
        if offset != UInt64(s.received) {
            reset(); return .drop("offset \(offset) != received \(s.received) for stream \(streamId)")
        }
        if s.received + data.count > s.buffer.count {
            reset(); return .drop("chunk overruns declared size for stream \(streamId)")
        }
        s.buffer.replaceSubrange(s.received ..< s.received + data.count, with: data)
        s.received += data.count
        lastProgress = now
        return .ok
    }

    func end(streamId: UInt64, totalSize: UInt64, sha256: Data, now: Date) -> StreamResult {
        guard let s = slots[streamId] else { return .ignore("unknown stream_id \(streamId)") }
        if s.done { reset(); return .drop("duplicate end for stream \(streamId)") }
        if totalSize != UInt64(s.buffer.count) || s.received != s.buffer.count {
            reset()
            return .drop("end size \(totalSize)/\(s.received) != declared \(s.buffer.count) for stream \(streamId)")
        }
        let actual = Data(SHA256.hash(data: s.buffer))
        if actual != sha256 { reset(); return .drop("hash mismatch for stream \(streamId)") }
        s.done = true
        lastProgress = now
        return .ok
    }

    /// The materialised item once every stream has verified, else nil.
    /// Consumes the pending state.
    func takeCompleted() -> ClipboardItem? {
        guard var item = pending else { return nil }
        if slots.values.contains(where: { !$0.done }) { return nil }
        item.formats = item.formats.map { f in
            if case .stream(let id) = f.payload, let s = slots[id] {
                return ClipFormat(mime: f.mime, size: UInt64(s.buffer.count), payload: .inline(s.buffer))
            }
            return f
        }
        reset()
        return item
    }

    func isStale(now: Date) -> Bool {
        pending != nil && now.timeIntervalSince(lastProgress) > Self.idleWindow
    }

    /// Drop the pending item if it has seen no progress within the window.
    /// Returns true if something was dropped.
    @discardableResult
    func dropStale(now: Date) -> Bool {
        guard isStale(now: now) else { return false }
        reset()
        return true
    }

    func reset() {
        pending = nil
        slots.removeAll()
    }
}
